### DISCUSSION TITLE
What is the process for running migrations to create tables when utilizing Exandra in a Phoenix application?

### DISCUSSION THREAD

**Discussion by Himanshu21git** [vinniefranco/exandra#45] [Q&A] [answered]
Hello @vinniefranco and @whatyouhide,

I hope you are both in good health.

I am relatively new to the Phoenix framework, with most of my prior experience being with relational databases like Postgres. However, I am now exploring the possibility of using a NoSQL database such as ScyllaDB for the backend of my application.

After reviewing the available documentation [here](https://hexdocs.pm/exandra/0.9.1/Exandra.html), I encountered some challenges when attempting to run migrations in my project, and I would greatly appreciate your assistance with this matter.

Here are the steps I have taken so far:

1. My **Elixir version is 1.16.0**, and I am using **Erlang/OTP 26**.
2. I initiated a new Phoenix project with the command `mix phx.new exandra_demo`.
3. I removed the `postgrex` dependency and added `exandra 0.9.1` to the project's dependencies. I then ran `mix deps.get` to update the project's dependencies.
4. As indicated in the documentation, I updated the repository configuration in **lib/exandra_demo/repo.ex** as follows:

```elixir
defmodule ExandraDemo.Repo do
  use Ecto.Repo,
    otp_app: :exandra_demo,
    adapter: Exandra
end
```

5. I added the database configuration in **config/dev.exs** as follows:

```elixir
# Configure your database
config :exandra_demo, ExandraDemo.Repo,
  cluster_name: :scylla_cluster,
  contact_points: ["127.0.0.1"], # list your Cassandra nodes here
  keyspace: "test_space",
  port: 9042, # default Cassandra port
  username: "",       # Replace with your username if applicable
  password: "",       # Replace with your password if applicable
  log: :info
```

6. After these configurations, I ran the `mix ecto.create` command and received a message in the terminal indicating that the database has already been created.

7. Subsequently, I attempted to create a context using the command `mix phx.gen.html Catalog Product products title:string description:string`, which generated the following schema and migrations:

**Schema: lib/exandra_demo/catalog/product.ex**

```elixir
defmodule ExandraDemo.Catalog.Product do
  use Ecto.Schema
  import Ecto.Changeset

  schema "products" do
    field :description, :string
    field :title, :string

    timestamps(type: :utc_datetime)
  end

  @doc false
  def changeset(product, attrs) do
    product
    |> cast(attrs, [:title, :description])
    |> validate_required([:title, :description])
  end
end
```

**Migration: priv/repo/migrations/20240112064913_create_products.exs**

```elixir
defmodule ExandraDemo.Repo.Migrations.CreateProducts do
  use Ecto.Migration

  def change do
    create table(:products) do
      add :title, :string
      add :description, :string

      timestamps(type: :utc_datetime)
    end
  end
end
```

9. However, when I attempt to run `mix ecto.migrate`, I encounter the following error:

```elixir
12:23:48.143 [error] Could not create schema migrations table. This error usually happens due to the following:

  * The database does not exist
  * The "schema_migrations" table, which Ecto uses for managing migrations, was defined by another library
  * There is a deadlock while migrating (such as using concurrent indexes with a migration_lock)

To fix the first issue, run "mix ecto.create" for the desired MIX_ENV.

To address the second, you can run "mix ecto.drop" followed by
"mix ecto.create," both for the desired MIX_ENV. Alternatively, you may
configure Ecto to use another table and/or repository for managing migrations:

    config :exandra_demo, ExandraDemo.Repo,
      migration_source: "some_other_table_for_schema_migrations",
      migration_repo: AnotherRepoForSchemaMigrations

The full error report is shown below.

** (Xandra.ConnectionError) action "checkout from cluster #PID<0.246.0>" failed with reason: not connected to any of the nodes
    (ecto_sql 3.11.1) lib/ecto/adapters/sql.ex:1054: Ecto.Adapters.SQL.raise_sql_call_error/1
    (elixir 1.16.0) lib/enum.ex:1700: Enum."-map/2-lists^map/1-1-"/2
    (ecto_sql 3.11.1) lib/ecto/adapters/sql.ex:1161: Ecto.Adapters.SQL.execute_ddl/4
    (ecto_sql 3.11.1) lib/ecto/migrator.ex:755: Ecto.Migrator.verbose_schema_migration/3
    (ecto_sql 3.11.1) lib/ecto/migrator.ex:563: Ecto.Migrator.lock_for_migrations/4
    (ecto_sql 3.11.1) lib/ecto/migrator.ex:432: Ecto.Migrator.run/4
    (ecto_sql 3.11.1) lib/ecto/migrator.ex:170: Ecto.Migrator.with_repo/3
    (ecto_sql 3.11.1) lib/mix/tasks/ecto.migrate.ex:151: anonymous fn/5 in Mix.Tasks.Ecto.Migrate.run/2
```

I kindly request your assistance in resolving this issue and would also appreciate it if you could review my project configurations to ensure it is correct. My goal is to run migrations and create a table in the database for further data querying and manipulation.

Thank you for your support.

**Comment by Himanshu21git** [comment, marked as answer]
After using [this fix](https://github.com/vinniefranco/exandra/issues/36#issuecomment-1696984879), the migrations were executed successfully.

**Comment by whatyouhide** [comment]
@Himanshu21git if you're willing to put in the time, we'd love a PR that improves documentation or error messages related to this 🙃

**Reply by Himanshu21git** [reply]
Sure, I would be happy to do that.

**Reply by Himanshu21git** [reply]
@whatyouhide 
Could you please let me know if there are any specific contribution guidelines that I should follow?

**Reply by whatyouhide** [reply]
Nothing specific! Just a PR and tests for the changes. 🙃

**Reply by Himanshu21git** [reply]
@whatyouhide  I have added the [PR](https://github.com/vinniefranco/exandra/pull/46) for it.
